### PR TITLE
dump list of applications data for single query

### PR DESCRIPTION
**Pull request by okerbez**

### CODE REVIEW

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java [hunk 1]
```diff
@@ -260,6 +265,17 @@ private <T> void callMethodWithReflection(
     }
   }
 
+  private static Optional<String> generateYarnApplicationJson(List<ImmutableMap<String, String> > listOfApplicationData) {
+    JSONArray jsonArray = new JSONArray();
+    listOfApplicationData.forEach(
```
Comment (ishmum123) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:270]: nit: would prefer `map` here as well. e.g. -

```
listOfApplicationData.stream().map(JSONObject::new).forEach(jsonArray::put);
```

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java [hunk 2]
```diff
@@ -223,6 +223,11 @@ public void postExecHook_recordsYarnApplicationDataWhenPossible() {
     report.setStartTime(100000L);
     report.setFinishTime(200000L);
     report.setFinalApplicationStatus(FinalApplicationStatus.UNDEFINED);
+    String expectedJson =
+        "[{\"YarnApplicationId\":\"application_1685098059769_1951\",\"HiveHostName\":\"test_host\",\"Queue\":\"test_queue\",\"YarnProcess\":\"10000.6\",\"YarnApplicationType\":\"TEZ\",\"YarnApplicationState\":\"RUNNING\",\"YarnDiagnostics\":\"test_diagnostics\",\"YarnCurrentApplicationAttemptId\":\"appattempt_1685098059769_1951_000001\",\"YarnUser\":\"test_user\",\"YarnStartTime\":\"100000\",\"YarnFinishTime\":\"200000\",\"YarnFinalApplicationStatus\":\"UNDEFINED\",\"YarnReportNumUsedContainers\":\"3\",\"YarnReportNumReservedContainers\":\"4\",\"YarnReportMemorySeconds\":\"100\",\"YarnReportVcoreSeconds\":\"400\",\"YarnReportQueueUsagePercentage\":\"90.5\",\"YarnReportClusterUsagePercentage\":\"64.5\",\"YarnReportPreemptedMemorySeconds\":\"200\",\"YarnReportPreemptedVcoreSeconds\":\"300\",\"YarnReportUsedResources\":\"<memory:800,"
```
Comment (ishmum123) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java:227]: Can we please have tests with 0 and 2 elements in the JSON array?

**Review by ishmum123 — CHANGES_REQUESTED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java [hunk 3]
```diff
@@ -55,27 +58,27 @@ public static Optional<ApplicationId> determineApplicationId(
...
-  private static Optional<ApplicationId> determineTezApplicationId() {
+  private static List<ApplicationId> determineTezApplicationId() {
     SessionState sessionState = SessionState.get();
     if (sessionState != null) {
       TezSessionState tezSessionState = sessionState.getTezSession();
       if (tezSessionState != null) {
         TezClient tezClient = tezSessionState.getSession();
         if (tezClient != null) {
-          return Optional.ofNullable(tezClient.getAppMasterApplicationId());
+          return Collections.singletonList(tezClient.getAppMasterApplicationId());
```
[the remark this answers is not in the document]
Comment (okerbez) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java]: done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java [hunk 4]
```diff
@@ -146,37 +147,38 @@ private GenericRecord getPostHookEvent(HookContext hookContext, EventStatus stat
...
-            applicationId -> {
-              recordBuilder.set("YarnApplicationId", applicationId.toString());
-              yarnApplicationRetriever
-                  .retrieve(conf, applicationId)
+            .forEach(applicationId -> {
+              ImmutableMap.Builder<String, String> applicationDataMap = ImmutableMap.builder();
+
+              applicationDataMap.put("YarnApplicationId", applicationId.toString());
+              yarnApplicationRetriever.retrieve(conf, applicationId)
                   .ifPresent(
```
[the remark this answers is not in the document]
Comment (okerbez) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:157]: ack

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java (hunk 1 above)
Comment (okerbez) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java]: done

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java (hunk 2 above)
Comment (okerbez) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java]: added test for 2 elements, will add 0 elements test in separate pull request

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java [hunk 5]
```diff
@@ -36,7 +39,7 @@ public class ApplicationIdRetriever {
 
   private static final Logger LOG = LoggerFactory.getLogger(ApplicationIdRetriever.class);
 
-  public static Optional<ApplicationId> determineApplicationId(
+  public static List<ApplicationId> determineApplicationId(
```
[the remark this answers is not in the document]
Comment (okerbez) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java]: done

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java (hunk 2 above)
Comment (okerbez) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java]: added test for 0 element case in this pull request

**Pull request merged**